File: nim/src/benchmarks/words.nim
import std/[math, strutils, tables]
import ../benchmark
import ../helper

type
  Words* = ref object of Benchmark
    words: int64
    wordLen: int64
    text: string
    checksumVal: uint32

proc newWords(): Benchmark =
  Words()

method name(self: Words): string = "Etc::Words"

method prepare(self: Words) =
  self.words = self.config_val("words")
  self.wordLen = self.config_val("word_len")

  let chars = "abcdefghijklmnopqrstuvwxyz"
  var wordsList: seq[string] = @[]

  for i in 0..<self.words.int:
    let len = nextInt(int32(self.wordLen)) + nextInt(3) + 3
    var word = newStringOfCap(len)
    for j in 0..<len:
      let idx = nextInt(chars.len.int32)
      word.add(chars[idx])
    wordsList.add(word)

  self.text = wordsList.join(" ")
  self.checksumVal = 0

method run(self: Words, iteration_id: int) =

  var frequencies = initCountTable[string]()

  for word in self.text.split(' '):
    if word.len == 0: continue
    frequencies.inc(word)

  var maxWord = ""
  var maxCount = 0

  for word, count in frequencies.pairs:
    if count > maxCount:
      maxCount = count
      maxWord = word

  let freqSize = uint32(frequencies.len)
  let wordChecksum = helper.checksum(maxWord)

  self.checksumVal += uint32(maxCount) + wordChecksum + freqSize

method checksum(self: Words): uint32 =
  return self.checksumVal

registerBenchmark("Etc::Words", newWords)

File: julia/src/benchmarks/Words.jl
mutable struct Words <: AbstractBenchmark
    words::Int64
    word_len::Int64
    text::String
    checksum_val::UInt32

    function Words()
        words = Helper.config_i64("Etc::Words", "words")
        word_len = Helper.config_i64("Etc::Words", "word_len")
        new(words, word_len, "", UInt32(0))
    end
end

name(b::Words)::String = "Etc::Words"

function prepare(b::Words)
    chars = 'a':'z'
    char_count = length(chars)

    words = b.words
    word_len = b.word_len

    text_parts = String[]
    sizehint!(text_parts, words)

    for i = 1:words
        len = Helper.next_int(word_len) + Helper.next_int(3) + 3
        word = String([chars[Helper.next_int(char_count)+1] for _ = 1:len])
        push!(text_parts, word)
    end

    b.text = join(text_parts, ' ')
end

function run(b::Words, iteration_id::Int64)

    frequencies = Dict{String,Int}()

    for word in split(b.text, ' ')
        if word == ""
            continue
        end
        frequencies[word] = Base.get(frequencies, word, 0) + 1
    end

    max_word = ""
    max_count = 0

    for (word, count) in frequencies
        if count > max_count
            max_count = count
            max_word = word
        end
    end

    freq_size = UInt32(length(frequencies))
    word_checksum = Helper.checksum(max_word)

    b.checksum_val += UInt32(max_count) + word_checksum + freq_size
end

function checksum(b::Words)::UInt32
    return b.checksum_val
end
